src/world/Spell/Spell.cpp

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright (c) 2014-2018 AscEmu Team <http://www.ascemu.org>
+This file is released under the MIT license. See README-MIT for more information.
+*/
+
+#include "Spell.h"
+#include "SpellAuras.h"
+#include "Definitions/AuraStates.h"
+
+SpellCastResult Spell::canCast(bool tolerate)
+{
+    if (p_caster != nullptr && !GetSpellInfo()->IsPassive() && !m_triggeredSpell)
+    {
+        // You can't cast other spells if you have the player flag preventing cast
+        if (p_caster->hasPlayerFlags(PLAYER_FLAG_PREVENT_SPELL_CAST))
+        {
+            return SPELL_FAILED_SPELL_IN_PROGRESS;
+        }
+
+        // Check for cooldown
+        if (!p_caster->Cooldown_CanCast(GetSpellInfo()))
+        {
+            if (m_triggeredByAura)
+                return SPELL_FAILED_DONT_REPORT;
+            else
+                return SPELL_FAILED_NOT_READY;
+        }
+    }
+
+    if (u_caster != nullptr)
+    {
+        bool requireCombat = true;
+        if (u_caster->hasAuraWithAuraEffect(SPELL_AURA_IGNORE_TARGET_AURA_STATE))
+        {
+            for (uint32_t i = MAX_TOTAL_AURAS_START; i < MAX_TOTAL_AURAS_END; ++i)
```

Comment:
Why is this a u32? Please use auto for implicit typing, or size_t to index arrays.